### ISSUE TITLE
HTML entity causes UnicodeEncodeError in LxmlLinkExtractor

### ISSUE THREAD

**Issue by fgpietersz** [scrapy/scrapy#998]
A page containing &amp;#9001; causes with error in LxmlLinkExtractor.link_extractor()

```
          File "/[path]/spiders/search_spider.py", line 112, in parse
            self.link_extractor.extract_links(response) if
          File "/[path to virtualenv]/local/lib/python2.7/site-packages/scrapy/contrib/linkextractors/lxmlhtml.py", line 107, in extract_links
            links = self._extract_links(doc, response.url, response.encoding, base_url)
          File "/[path to virtualenv]/local/lib/python2.7/site-packages/scrapy/linkextractor.py", line 94, in _extract_links
            return self.link_extractor._extract_links(*args, **kwargs)
          File "/[path to virtualenv]/local/lib/python2.7/site-packages/scrapy/contrib/linkextractors/lxmlhtml.py", line 57, in _extract_links
            url = url.encode(response_encoding)
          File "/[path to virtualenv]/lib/python2.7/encodings/cp1252.py", line 12, in encode
            return codecs.charmap_encode(input,errors,encoding_table)
        exceptions.UnicodeEncodeError: 'charmap' codec can't encode character u'\u2329' in position 87: character maps to <undefined>
```

The encoding is correct and works for displaying the page in a browser, but fails as above in Scrapy.

**Comment by Digenis** [comment]
Here is a snippet to reproduce:
file: snowman.html
```html
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=Windows-1251"/>
  </head>
  <body>
    <a title="unicode snowman" href="./&#9731;">
      &#x0441;&#x043d;&#x0435;&#x0436;&#x0435;&#x043d;
      &#x0447;&#x043e;&#x0432;&#x0435;&#x043a;
    </a>
  </body>
</html>
```
~~run: `scrapy shell ~/snowman.html`~~
The above doesn't work (probably some bug `utils.url`),
neither does `scrapy shell file:///home/digenis/snowman.html`.
It takes 2 slashes and a dummy step:
`scrapy shell file://whatever/home/digenis/snowman.html`

```python
from scrapy.linkextractors import LinkExtractor
LinkExtractor().extract_links(response)
```
 `UnicodeEncodeError: 'charmap' codec can't encode character u'\u2603' in position 24: character maps to <undefined>`

lxml decodes the entity inside the href attribute
and since`Windows-1251` doesn't have the :snowman: character
it can't be decoded "back" to cp1251.

As @kmike suggests in https://github.com/scrapy/scrapy/issues/1403#issuecomment-149194449
the path part needs to be encoded to utf8, as browsers do.
It should need to be able to be encoded in the response's encoding

**Comment by redapple** [comment]
This issue is fixed in #1949 where Link objects are created with Unicode URLs (and [not converted to native str anymore](https://github.com/scrapy/scrapy/pull/1949/files#diff-952f1c77a1cfacb08eed58f08daa8870L59), which chokes in Python 2),
leaving the encoding to the Request' init (when safe_url_string is called, path encoded as UTF-8 and query/fragment as response/document encoding)
